File: cpp/761_make_largest_special/make_largest_special.cpp
#include <algorithm>
#include <cassert>
#include <numeric>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  string makeLargestSpecial(string s) {
    if (s.size() <= 2) {
      return s;
    }
    int cnt = 0, left = 0;
    vector<string> subs;
    for (int i = 0; i < s.size(); ++i) {
      if (s[i] == '1') {
        ++cnt;
      } else {
        --cnt;
        if (cnt == 0) {
          subs.push_back("1" + makeLargestSpecial(s.substr(left + 1, i - left - 1)) + "0");
          left = i + 1;
        }
      }
    }

    sort(subs.begin(), subs.end(), greater<string>{});
    string ans = accumulate(subs.begin(), subs.end(), ""s);
    return ans;
  }
};

int main() {
  vector<tuple<string, string>> tests{
      {"11011000", "11100100"},
      {"10", "10"},
  };

  for (auto& [s, ans] : tests) {
    assert(Solution().makeLargestSpecial(s) == ans);
  }
  return 0;
}

File: golang/761_make_largest_special/make_largest_special.go
// Package main ...
package main

import (
	"sort"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func makeLargestSpecial(s string) string {
	if len(s) <= 2 {
		return s
	}
	subs := sort.StringSlice{}
	cnt, left := 0, 0
	for i, ch := range s {
		if ch == '1' {
			cnt++
		} else if cnt--; cnt == 0 {
			subs = append(subs, "1"+makeLargestSpecial(s[left+1:i])+"0")
			left = i + 1
		}
	}
	sort.Sort(sort.Reverse(subs))
	return strings.Join(subs, "")
}

func main() {
	tests := []struct {
		s   string
		ans string
	}{
		{"11011000", "11100100"},
		{"10", "10"},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, makeLargestSpecial(test.s), index)
	}
}
